### DISCUSSION TITLE
Rounded custom clip shapes

### DISCUSSION THREAD

**Discussion by erictoader** [varabyte/kobweb#365] [Q&A]
My goal is to create a trapezoid with rounded edges.  I was able to cut the shape right but it seems that rounding it does not work as fluently (and with good reason, this can only be done via a Gaussian Blur filter on an SVG)

This CSS code would create an SVG with rounded edges and clip it in a trapezoid shape.
```css
<svg width="200" height="200" viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg">
    <defs>
        <filter id="rounded">
            <feGaussianBlur in="SourceGraphic" stdDeviation="3" result="blur" />
            <feColorMatrix in="blur" mode="matrix" 
                           values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 19 -9" 
                           result="rounded" />
        </filter>
    </defs>
    <polygon points="0,0 200,0 120,200 0,200" fill="#3498db" filter="url(#rounded)" />
</svg>
```

I'm not very familiar with CSS, this code was spat out by ChatGPT, but several SO posts seem to indicate you have to do this in order to have rounded borders on a clip shape.

This is my current code:

```kt
@Composable
fun Highlight() {
    Box(
        modifier = Modifier
            .width(200.px)
            .height(300.px)
            .clip(IrregularRectangle())
            .borderRadius(20.px)
            .backgroundColor(Color.red)
    )
}

class IrregularRectangle : Shape {
    override val path: Path
        get() = PolygonPath(
            Pair(0f, 0f),
            Pair(100f, 0f),
            Pair(60f, 100f),
            Pair(0f, 100f)
        )
}
```

This is my current output:
<img width="288" alt="Screenshot 2023-10-23 at 18 25 00" src="https://github.com/varabyte/kobweb/assets/99261319/0b8dcc68-9a94-4620-9d20-bd3eb4e53cbf">

This is my desired output (I'm not good at drawing; the point is that I want all of the edges to be rounded):
<img width="280" alt="Screenshot 2023-10-23 at 18 25 07" src="https://github.com/varabyte/kobweb/assets/99261319/6958a601-cd1e-40ec-a2c9-039e5d153a50">

It seems that the Svg builder would need a Filter {} builder in order to invoke that feGaussianBlur stuff.

<h2>
If you are aware of a method that this can be achieved right now, please let me know.</br>
I only picked this framework up a day ago :D
</h2>

**Comment by bitspittle** [comment]
I moved this to discussions in case people have other suggestions, but I opened #366 as a feature request in response to this question.

I think what's most likely though is I just need to add those missing bits for you. Depending on your skill level / courage, you could jump into [the SVG code](https://github.com/varabyte/kobweb/blob/main/frontend/compose-html-ext/src/jsMain/kotlin/com/varabyte/kobweb/compose/dom/svg/Svg.kt) and see how it's working, extending the extra bits yourself locally so you can be unblocked faster.

**Comment by bitspittle** [comment]
Just an FYI, this is *not* checked in yet, but I'm actively working on SVG filters right now. Hopefully will have more updates later today / tomorrow. It will be a at least two weeks before 0.14.3 is released, but I will push out an early snapshot in the next day or two.

The following code works for me *locally*:

```kotlin
Svg(attrs = {
    viewBox(0, 0, 480, 200)
}) {
    val blurMe = SvgId("blurMe")
    Defs {
        Filter(id = blurMe) {
            GaussianBlur {
                in1(SVGFEInput.SourceGraphic)
                stdDeviation(5)
            }
        }
    }

    Circle {
        cx(60)
        cy(60)
        r(50)
        fill(Colors.Green)
    }

    Circle {
        cx(170)
        cy(60)
        r(50)
        fill(Colors.Green)
        filter(blurMe)
    }
}
```

which was adapted from https://developer.mozilla.org/en-US/docs/Web/SVG/Element/filter#example

**Reply by bitspittle** [reply]
**Update** @erictoader 

The following code, which handles your case, will be supported in 0.14.3:

```kotlin
Svg(attrs = {
    width(200)
    height(200)
    viewBox(0, 0, 200, 200)
}) {
    val roundedId = SvgId("rounded")
    Defs {
        Filter(roundedId) {
            GaussianBlur {
                `in`(SVGFEInput.SourceGraphic)
                stdDeviation(3)
                result("blur")
            }
            ColorMatrix {
                `in`("blur")
                type(SVGFEColorMatrixType.Matrix)
                values(
                    1, 0, 0, 0, 0,
                    0, 1, 0, 0, 0,
                    0, 0, 1, 0, 0,
                    0, 0, 0, 19, -9,
                )
            }
        }
    }

    Polygon {
        points(0 to 0, 200 to 0, 120 to 200, 0 to 200)
        fill(Color.rgb(0x3498db))
        filter(roundedId)
    }
}
```
![trapezoid-with-rounded-corners](https://github.com/varabyte/kobweb/assets/6038046/9fa15c1f-90c5-442f-b1f1-69d83b312a6d)

If you want to try this yourself *right now*, go into your project's `libs.versions.toml` file and change the Kobweb version to `0.14.3-SNAPSHOT`. 

> [!WARNING]
> Using snapshots are living on the bleeding edge a little, but it's usually pretty safe to do, and you can ping me if I break you.
